File: cPP_java_py_js-DSA/1ARRAY_STRINGS/palindrome/palindrome.js
import { printComplexityOfSolution } from "../../0commonFunctions/printComplexities.js";
// Problem Statement : Given a string and we have to check whether it is a palindrome or not.

let str = "naman";
let str2 = "suresh";
function checkPalindrome(str) {
  // concept : two pointer ,comparison
  // time :O(n) , space : O(1)
  printComplexityOfSolution("Optimised Two Pointer", "O(N)", "O(1)");
  let i = 0,
    j = str.length - 1;
  while (i < j) {
    if (str[i] != str[j]) return false;
    i++;
    j--;
  }
  return true;
}
console.log("check palindrome ", str, checkPalindrome(str));
console.log("check palindrome ", str2, checkPalindrome(str2));

File: cPP_java_py_js-DSA/1ARRAY_STRINGS/palindrome/palindrome.cpp
// check palindrome - cPP code is pending
#include <bits/stdc++.h>
#include "../../0commonFunctions/printComplexity.h"
using namespace std;
bool checkPalindromeString_Approach_Optimised(string s)
{

    // concept : two pointer + comparison
    // time : O(n), space :O(1)
    printComplexities("O(n)", "O(1)");
    int st = 0, en = s.size() - 1;
    while (st < en)
    {
        if (s[st] != s[en])
            return false;
        st++;
        en--;
    }
    return true;
}
int main()
{
    string s = "naman";
    string s2 = "raman";
    cout << checkPalindromeString_Approach_Optimised(s) << endl;
    cout << checkPalindromeString_Approach_Optimised(s2) << endl;

    return 0;
}